contracts/scroll/ScrollVerifierHooks.sol

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IVerifierHooks, InvalidProof, NOT_A_CONTRACT, NULL_CODE_HASH} from '../IVerifierHooks.sol';
+
+interface IPoseidon {
+    function poseidon(
+        uint256[2] memory,
+        uint256
+    ) external view returns (bytes32);
+}
+
+//import "forge-std/console2.sol";
+
+contract ScrollVerifierHooks is IVerifierHooks {
+    IPoseidon immutable _poseidon;
+    constructor(IPoseidon poseidon) {
+        _poseidon = poseidon;
+    }
+
+    // https://github.com/scroll-tech/scroll/blob/738c85759d0248c005469972a49fc983b031ff1c/contracts/src/libraries/verifier/ZkTrieVerifier.sol#L259
+    // https://github.com/scroll-tech/go-ethereum/blob/staging/trie/zk_trie.go#L176
+    // https://github.com/scroll-tech/zktrie/blob/main/trie/zk_trie_proof.go#L30
+    // https://github.com/ethereum/go-ethereum/blob/master/trie/proof.go#L114
+    // https://github.com/scroll-tech/mpt-circuit/blob/v0.7/spec/mpt-proof.md#storage-segmenttypes
+
+    // 20240622
+    // we no longer care (verify or require) about the magic bytes, as it doesn't do anything
+    // https://github.com/scroll-tech/zktrie/blob/23181f209e94137f74337b150179aeb80c72e7c8/trie/zk_trie_proof.go#L13
+    // bytes32 constant MAGIC = keccak256("THIS IS SOME MAGIC BYTES FOR SMT m1rRXgP2xpDI");
+
+    // https://github.com/scroll-tech/zktrie/blob/23181f209e94137f74337b150179aeb80c72e7c8/trie/zk_trie_node.go#L30
+    uint256 constant NODE_LEAF = 4;
+    uint256 constant NODE_LEAF_EMPTY = 5;
+    uint256 constant NODE_LEAF_LEAF = 6; // XX
+    uint256 constant NODE_LEAF_BRANCH = 7; // XB
+    uint256 constant NODE_BRANCH_LEAF = 8; // BX
+    uint256 constant NODE_BRANCH_BRANCH = 9; // BB
+
+    // 20240918: 900k gas
+    function verifyAccountState(
+        bytes32 stateRoot,
+        address account,
+        bytes memory encodedProof
+    ) external view returns (bytes32 storageRoot) {
+        bytes[] memory proof = abi.decode(encodedProof, (bytes[]));
+        //bytes32 raw = bytes32(bytes20(account));
+        bytes32 key = poseidonHash1(bytes20(account)); // left aligned
+        (bytes32 leafHash, bytes memory leaf) = walkTree(key, proof, stateRoot);
+        // HOW DO I TELL THIS DOESNT EXIST?
```

Comment:
:) Im cognisant of the fact that comments like this are indicative that there might be non-working edge cases..?

Reply:
The entire `ScrollVerifierHooks` was written with a ton of debugging as I was trying to decipher what they we're doing.  The first realization was the magic bytes does nothing.  The second was the lack of data to prove exclusion.  Any comments were related to simulating it with various "proof of zero" examples and attempting to figure out what was going on.  My conclusion was that the current proof data is insufficient.  The code is still in that state minus print statements.